x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationService.java

```diff
@@ -462,11 +462,15 @@ private void consumeToken(AuthenticationToken token) {
                                 // the user was not authenticated, call this so we can audit the correct event
                                 request.realmAuthenticationFailed(authenticationToken, realm.name());
                                 if (result.getStatus() == AuthenticationResult.Status.TERMINATE) {
-                                    logger.info("Authentication of [{}] was terminated by realm [{}] - {}",
-                                        authenticationToken.principal(), realm.name(), result.getMessage());
-                                    Exception e = (result.getException() != null) ? result.getException()
-                                        : Exceptions.authenticationError(result.getMessage());
-                                    userListener.onFailure(e);
+                                    if (result.getException() != null) {
+                                        logger.info(new ParameterizedMessage("Authentication of [{}] was exceptionally terminated by " +
+                                                        "realm [{}] - {}", authenticationToken.principal(), realm.name(),
+                                                        result.getMessage()), result.getException());
+                                    } else {
+                                        logger.info("Authentication of [{}] was gracefully terminated by realm [{}] - {}",
+                                                authenticationToken.principal(), realm.name(), result.getMessage());
+                                    }
+                                    userListener.onFailure(result.getException());
```

Comment:
I didn't test, but the same goal could be achieved by calling `userListener.onResponse(null)` in the `else` block, i.e.:
```java
if (result.getException() != null) {
    ...
} else {
   ...
   userListener.onResponse(null);
}
```
This leverages the existing logic inside `consumeUser` so that changes to the following `IteratingActionListener`'s error handler is not needed. I personally feels this would be simpler and easy to reason about.

Reply:
Nope. If you call `userListener.onResponse(null)` the `IteratingActionListener` goes through on to the next realm in the chain, which is exactly what `AuthenticationResult.Status.TERMINATE` should prevent.

Reply:
You are absolutely correct. Sorry I mis-read the logic.